src/Abp/Runtime/Caching/AbpCacheBase.cs

```diff
@@ -114,16 +114,14 @@ public virtual TValue[] Get(TKey[] keys, Func<TKey, TValue> factory)
                     var fetched = new List<KeyValuePair<TKey, TValue>>();
                     for (var i = 0; i < items.Length; i++)
                     {
-                        TKey key = keys[i];
-                        TValue value = items[i];
-                        if (value == null)
+                        if (items[i] == null)
                         {
-                            value = factory(key);
-                        }
-
-                        if (value != null)
-                        {
-                            fetched.Add(new KeyValuePair<TKey, TValue>(key, value));
+                            var key = keys[i];
```

Comment:
Why change to not assign `value` to a variable but still assign `key` to a variable?

Could be a one-liner change instead:

```diff
  if (value != null)
  {
+     items[i] = value;
      fetched.Add(new KeyValuePair<TKey, TValue>(key, value));
  }
```

Reply:
hi 
If the `items[i]` is not generated by a factory method, we should not call the `Set `method.

Reply:
> If the items[i] is not generated by a factory method, we should not call the Set method.
 
currently both `Get` single and `Get` multiple call `Set` whether the value is generated by factory or not. 

If we want to fix this behavior, then we need to fix it for both single key and multiple keys m`Get`

Reply:
hi @ryancyq 

In the ` Get single` method, the factory method and the Set method are called only when the `items == null`. It is not the same as the `Get multiple` method.

https://github.com/aspnetboilerplate/aspnetboilerplate/pull/5277/files#diff-c06cc5b3081a05d70f8667811468c187R59-R76

Reply:
I see. in that case how about combining what @acjh suggested with the following

```csharp
TKey key = keys[i];
TValue value = items[i];
if (value == null)
{
    items[i] = factory(key);
}
else
{
    fetched.Add(new KeyValuePair<TKey, TValue>(key, value));
}
```

Reply:
i see. In that case, how about combined what @acjh  suggested with the followings:

```csharp
TKey key = keys[i];
TValue value = items[i];
if (value == null)
{
    value = factory(key);
    items[i] = value;
}
else
{
    fetched.Add(new KeyValuePair<TKey, TValue>(key, value));
}
```

Reply:
@maliming I see.

@ryancyq Still not quite right, need to check if `value` is `null` after `value = factory(key);`.

In this case, we can keep @maliming's change with the following changes:

```diff
  var key = keys[i];
- items[i] = factory(key);
- if (items[i] != null)
+ var item = items[i] = factory(key);
+ if (item != null)
  {
-     fetched.Add(new KeyValuePair<TKey, TValue>(key, items[i]));
+     fetched.Add(new KeyValuePair<TKey, TValue>(key, item));
  }
```

Better yet, do early return:

```c#
if (items[i] != null)
{
    continue;
}

var key = keys[i];
var item = items[i] = factory(key);
if (item == null)
{
    continue;
}

fetched.Add(new KeyValuePair<TKey, TValue>(key, item));
```

Reply:
@acjh this following doesn't seem to work as we still unable to determine the value generated by the factory
```diff
var key = keys[i];
- items[i] = factory(key);
- if (items[i] != null)
+ var item = items[i] = factory(key);
+ if (item != null)
  {
-     fetched.Add(new KeyValuePair<TKey, TValue>(key, items[i]));
+     fetched.Add(new KeyValuePair<TKey, TValue>(key, item));
  }
```

@maliming how about
```csharp
TKey key = keys[i];
TValue value = null;
if (items[i] == null)
{
    value = factory(key);
    items[i] = value;
}

if (value != null)
{
    fetched.Add(new KeyValuePair<TKey, TValue>(key, value));
}
```

Reply:
What do you mean by "unable to determine the value generated by the factory"?
`item` is the value generated by the factory.

Reply:
`var item = items[i] = factory(key);` by assigning the factory generated value into `item`, we lost the information of whether `item` is filled up by `items[i]` or `factory(key)`

Reply:
How is there any doubt that `item` is filled with the value of `factory(key)`?

```c#
int[] items = {0};
Console.WriteLine(items[0]); // 0

var item = items[0] = 1;
Console.WriteLine(item);     // 1
Console.WriteLine(items[0]); // 1
```

Perhaps you missed that the first code block in https://github.com/aspnetboilerplate/aspnetboilerplate/pull/5277#r377135879 is implied to be wrapped in `if (items[i] == null)` as I specified that "we can keep **@<span></span>maliming**'s change" with those changes, i.e.

```c#
if (items[i] == null)
{
    var key = keys[i];
    var item = items[i] = factory(key);
    if (item != null)
    {
        fetched.Add(new KeyValuePair<TKey, TValue>(key, item));
    }
}
```

The second code block is not.

Reply:
@acjh i added some commits to minimise the code change. Let me know what you think, i can revert to @maliming changes if needed.